PixelPaint: Specialize how the mouse position is converted into a pixel position based on the currently selected tool

Crax97 opened the pull request:
This commit aims to fix the issue #15064, which happens because the `event_with_pan_and_scale_applied` has been modified to improve the feeling of the rectangle tool.
The issue is that now a pixel offset has been applied because all the other tools work best when the input position is floored instead of being rounded: to fix this situation, the `Tool::point_position_to_preferred_cell `virtual method has been added which is used to fetch the pixel that a tool should interact with based on the mouse position.
The `RectangleSelectionTool `specializes this method to improve how the selection is expanded.

GMTA commented on Userland/Libraries/LibGfx/Point.h:
```diff
@@ -246,6 +246,12 @@ class Point {
+    [[nodiscard]] Point<U> to_floored() const
+    {
+        return Point<U>(floorf(x()), floorf(y()));
+    }
```
Usage of `floorf` assumes `U` is `float`; usually `.to_type<int>()` is used instead if all values are positive.

Crax97 replied:
I see, in that case i have some questions:
1. Since it seems like the coordinates in the image editor are expressed in the range from [0, 0] to [width,height] (correct me if i'm wrong), to fix the issue i could simply use `to_type<int>()`, right? 
2. In such case, wouldn't that be like reverting the change Andreas made in e6d860f2fe49ad8f39300087d1d9da3e052f3faf? And should i remove the to_floored function then? I assume there may be situations where flooring a point would be needed in the future

GMTA replied:
1: yes
2: yes, and yes: we shouldn't add functions just because they _could_ be used, they should actively be used in the PR

You should doublecheck whether this PR isn't breaking the very thing Andreas has tried to fix, the very least :-)

Crax97 commented:
The latest 5 commits are the only ones pertaining to the PR, i am sorry for this absolute mess!

awesomekling commented:
The rectangle select tool is much worse after these changes. Please retain its rounding behavior, and change only the behavior of other tools.

awesomekling commented:
Also, please squash "Improve robustness of Point::to_ceiled" into the commit that added it instead of having a separate tweak commit :)

Crax97 commented:
I have fixed all the things mentioned in the comments.
While i still think that a pixel should selected as soon as the cursor is on it (and not when the cursor is over it's center), i reverted those changes too and kept the rectangle selection tool's rounding behavior
